Pointsクラス作成・マジックナンバーの定数化

mkdirman opened the pull request:
・Pointクラス作成
・マジックナンバーの削除にあたって、Position,Size,StoneColorクラスを新たに作成

Commit:
・GameFactory,PlayersFactoryの追加。
・Turn,Modeクラスに入力→チェックの機能をそれぞれに持たせた。

Commit:
・Mode,Turnクラスを消去しModeGame,ModeTurnクラス(Enum)に統合。
・GameFactoryをcreateのみに修正→main関数内に変更。
・

Commit:
・GameFactoryのcreateの分岐にelse:ValueErrorを追加。

Commit:
marge競合

Commit:
・Pointクラス作成
・マジックナンバーの削除に伴い、Position,Size,StoneColorクラスを新たに作成。

hagikento commented on osero_kadai/models/reversi_board.py:
```diff
@@ -2,10 +2,36 @@
+    black = 1
+    white = -1
+
 class ReversiBoard():
```
class ReversiBord:
  ROW_NUM = 8
  COLUMN_NUM = 8
   ...
のような形でクラスの定数を宣言する(処理の中で値を変えないものは定数にする)

hagikento commented on osero_kadai/models/reversi_board.py:
```diff
@@ -2,192 +2,234 @@
+        y_on_square = Size.convert_to_square(y)
+        x_on_square = Size.convert_to_square(x)
+
+        if (x == Position.row_number)|(y == Position.column_number):
```
if point.is_壁:

hagikento replied:
class Point:
    def __init__(self, x: int, y: int)->None:
        if x < 1 or 8 < x:
            raise InvalidInputValueException
        if y < 1 or 8 < y:
            raise InvalidInputValueException
        self.x = x
        self.y = y


    def is_壁(self, row_num: int, column_num: int):
        return (self.x == row_num)|(self.y == column_num)


input = Point(-1, 2)


def input_point()->Point:
    while True:
        try:
            x = int(input('x座標を入力してね(1-8):'))
            y = int(input('y座標を入力してね(1-8):'))
            return Point(x, y)
        except InvalidInputValueException:
            print("再度入力してください")

Commit:
Pointクラス(仮)
・Point の__init__で行う範囲チェックにおいて、reversi_boardのwrite_number(行数と列数の表記を行う関数)では、行数列数を表記する場所が駒を置ける部分の範囲外の為、範囲チェックにひっかかる。そのため、Point(x,y)でまとめて値を受け渡すことが出来ない。
・available_listの意味合い的にはPointクラスに入れるのが良いと思われるが、listを作成する際にreversi_boardクラスのboardの情報等が必要で判断が難しい。

Commit:
Pointクラスにて、
・self.colorの追加
・行数列数の表示にも使えるように、__init__のraiseを変更。

Commit:
修正

hagikento commented on osero_kadai/models/reversi_board.py:
```diff
@@ -1,193 +1,236 @@
-        self.update_diagonal2(x, y, color)
+    def is_in_available_list(self, point: Point):
+        for point_available in self.available_list:
+            if (point_available.x == point.x)&(point_available.y == point.y):
```
ここのequeal判定とか
下記のようにインスタンス同士の比較にしたい
if point_available == point:

hagikento replied:
そうするにあたり、
Pointクラスにおいて
```
 def __eq__(self, other: Point):
     return self.x == other.x and self.y == other.y
```
みたいなのを作る必要あり

hagikento replied:
```
def is_same_color(self, other):
   return self.x == other.x and self.y == other.y and self.color == other.color


if a_point.is_same_color(b_point)

```
上記みたいなものもあっても良さそう

hagikento commented on osero_kadai/models/point.py:
```diff
@@ -0,0 +1,49 @@
+                print("再度入力してください")
+
+    @staticmethod
+    def input_board_init(board: np.array((9, 9))):
```
これはboardクラスが持つべきもの

hagikento commented on osero_kadai/models/reversi_board.py:
```diff
@@ -1,193 +1,236 @@
                 break
 
+    def is_the_point_has_0(self, point: Point):
+        return self.board[point.x, point.y] == 0
```
class board:

  elif is_右上_point_0(base_point, i):

  
def get_right_point()

def is_右上_point_0(base_point, i) -> Point:
     p = get_右上_point(base_point, i)
     retunr p.is_0

def get_右上_point(base_point, i) -> Point:
    p = base_point.get_右上_point(i)
    return self.get_point(p)

def get_point(point) -> Point:
   return self.board[point.x, point.y]

def get_point(x, y) -> Point:
  return self.board[x, y]

def is_point_0(x, y) -> bool:
   return self.get_point(x, y).is_0

Class Point
   def is_0(self) -> bool:
       return self.color == 0

def get_right_point(self, i) -> Point:
      return Point(self.x + i, self.y)

def get_右上_point(self, i) -> Point:
    return Point(self.x + i, self.y + i)

hagikento commented on osero_kadai/models/reversi_board.py:
```diff
@@ -1,193 +1,236 @@
+        return abs((self.board-board_before_update).sum()) != 1
+
+    def check_frip_over(self, point: Point):
+        if self.is_frip_over(point) == False:
```
```
if self.is_frip_over(point) is Flase:
if not self.is_frip_over(point):
```

Commit:
・boardをnp配列からPointクラスの二次のリストに変更。
→それに伴いreversi_boardの随所を変更。

問題点
・is_frip_overにて、np配列のコピーからクラスのリストのコピーに変わったため、コピーの情報量が増えた。

・is_update_boardにて、ボードの変化前と変化後を比べる仕組みがfor文に変更されたことで動作が遅くなった。
→使用回数が多い為、ここの仕組みは考え直す必要がある。

・Pointクラスをget_point()などで制御することにより、関数の数が膨大になった。

Commit:
reversi_boardにおいて、
・is_update_board()の仕組みの変更(ひっくりかえした数をカウント)
を行いました。

hagikento commented on osero_kadai/models/point.py:
```diff
@@ -0,0 +1,75 @@
+        return (self.x == 0)|(self.y == 0)
+
+    @property
+    def is_0(self):
```
is_0はメソッド名だけ見て意図が分かりずらい

is_same_colorのように、どのような意味のチェックなのかがわかるメソッド名にしたい

is_not_placed みたいな感じかな？
